mem/buffer_slice.go

```diff
@@ -150,6 +153,14 @@ func (r *sliceReader) Remaining() int {
 	return r.len
 }
 
+func (r *sliceReader) Reset(s BufferSlice) {
+	r.Close()
+	s.Ref()
+	r.data = s
+	r.len = s.Len()
+	r.bufferIdx = 0
```

Comment:
You could optimize very slightly by not calling `r.Close()` and just calling `r.data.Free()` instead, since you are initializing every field here.

But...that makes me wonder: why does `Close` bother clearing `data` and `len` in the first place?  I am thinking that's done intentionally "in case a read happens after `Close`".  If so, then are we sure it is okay to re-use them?

cc @PapaCharlie in case he remembers the motivation for this and can comment on the feasibility about reusing the readers here.

Reply:
Changed to avoid redundant work in `Reset`.  The godoc for `Close` mentions that it's fine for readers to call `Read` after close.
> Subsequent calls to Read will return (0, io.EOF).

If `Read` want' allowed after `Close` we could have pooled the `Reader` objects, putting object back in the poll at the end of `Close`. With the `Reset` method, the caller needs to decide when to call `Reset`.  If they plan to call `Read` from the old buffer slice afterwards, they should not call `Reset`.

Reply:
That makes sense.  We might have some places that are reading after Close, but presumably the control buf isn't one of them.